File: packages/widget/src/ui/identity-header.ts
// Who the customer is talking to, in the panel header.
//
// ── The two semantics this file exists to get right ──────────────────────
//
// `ChatSession.handledBy` (core, state/types.ts) carries exactly one bit of
// presentation information, and getting either half of it wrong tells the
// customer something false:
//
//   1. ABSENT means "render your own configured title" — never "nobody is
//      handling this chat". `status`/`mode` carry that signal, not this
//      field. An absent `handledBy` must never blank the header, spin it, or
//      say "no agent".
//
//   2. PRESENT does not always mean CURRENT. A session reactivated from
//      CLOSED/RESOLVED keeps its previous agent server-side, so `handledBy`
//      can still name someone while `status` is back to
//      `WAITING_FOR_AGENT` — that person is not actually on the chat right
//      now. Core's `isHandledByCurrent` (client/session.ts, T10) is the one
//      canonical gate for this, "same spirit as `deriveTickState`" per its
//      own doc — every binding calls it rather than re-deriving the
//      WAITING_FOR_AGENT special case locally. Both rules collapse into one
//      call: `isHandledByCurrent` already returns `false` for an absent
//      `handledBy` (its first conjunct), so gating the whole label on it
//      handles rule 1 and rule 2 with a single branch — there is no second,
//      separate "is it absent" check to keep in sync with it.
//
// `HandledBy`/`ChatStatus` are imported straight from `@dhaam-ccrm/core`
// rather than `@dhaam-ccrm/js`, matching message-list.ts's existing
// AttachmentMetadata/CloseReason workaround: the binding package re-exports
// most of ChatState's shape but not every type reachable from it, and a
// type-only import costs nothing at runtime.

import { isHandledByCurrent } from '@dhaam-ccrm/core';
import type { ChatSession } from '@dhaam-ccrm/js';

import { el } from './dom.js';

export interface IdentityHeaderView {
  /**
   * The title element itself — an `<h2 id="dh-title">`, the same id the
   * panel's `aria-labelledby` already points at. Mount this node where a
   * hand-built title `<h2>` used to go; do not create a second one.
   */
  readonly node: HTMLHeadingElement;

  /**
   * The announcement channel for identity CHANGES only. Separate from any
   * other live region in the panel (same reasoning as message-list.ts's
   * `liveRegion` split from its `log`): marking `node` itself live would
   * re-announce the title on every unrelated header re-render, and folding
   * this into the message log's region would race whatever it is announcing
   * at the same moment. Mount it once, anywhere in the panel — it carries no
   * visible layout.
   */
  readonly liveRegion: HTMLElement;

  /**
   * Replaces the title used when no agent is driving it — published config,
   * which arrives after mount. No-ops visually while an agent's name is
   * showing, and never announces.
   */
  setFallbackTitle(title: string): void;

  /**
   * Recomputes the displayed identity from `session.status`/`handledBy` and
   * speaks the change — but only when the DISPLAYED LABEL actually differs
   * from what was last shown, and never on the very first call.
   */
  update(session: Pick<ChatSession, 'status' | 'handledBy'> | null): void;

  /**
   * Explicitly sets the active title (e.g. store name when chatting with a merchant,
   * or Support when general platform chat).
   */
  setTitle(title: string): void;
}

/**
 * Checks if a title is a generic system title or a specific merchant/store title.
 */
function isGenericTitle(t: string | undefined): boolean {
  if (!t) return true;
  const lower = t.trim().toLowerCase();
  return (
    lower === 'chat with us' ||
    lower === 'admin support chat' ||
    lower === 'store support & chat' ||
    lower === 'dhaam ai' ||
    lower === 'support' ||
    lower === 'general support'
  );
}

/**
 * The platform's own name — what a generically-titled chat's agent label is
 * paired with once someone answers it ("Ravi · Dhaam Support"). Never used
 * for a SPECIFIC title (a store or merchant name) — that title stands alone,
 * unpaired with anyone; see {@link isGenericTitle} and {@link labelFor}.
 */
const PLATFORM_BRAND = 'Dhaam Support';

/**
 * @param initialTitle The widget's own configured title (`WidgetConfig.title`),
 *   shown whenever there is no CURRENT handler to name.
 */
export function createIdentityHeader(initialTitle: string): IdentityHeaderView {
  let fallbackTitle = initialTitle;
  const node = el('h2', { attrs: { class: 'dh-title', id: 'dh-title' }, text: fallbackTitle });

  const liveRegion = el('div', {
    attrs: { class: 'dh-sr', role: 'status', 'aria-live': 'polite', 'aria-atomic': 'true' },
  });

  let currentLabel = fallbackTitle;
  let seenAnyState = false;

  function currentAgentName(session: Pick<ChatSession, 'status' | 'handledBy'> | null): string | null {
    if (session !== null && isHandledByCurrent(session) && session.handledBy?.displayName) {
      return session.handledBy.displayName;
    }
    return null;
  }

  /**
   * A SPECIFIC title — a real store/outlet/merchant name a host configured on
   * purpose — names who this conversation is with all by itself. Pairing it
   * with whichever staff member happens to answer ("tse · Mohali") answers a
   * question the customer never asked and outranks the one fact that
   * actually matters here: which outlet they are talking to. So a specific
   * title is never touched by who is currently handling the chat; it is the
   * label, full stop.
   *
   * A GENERIC title (the widget's own default, or none) has the opposite
   * problem: "Chat with us" names no one, so once someone answers, pairing
   * their name with the platform's own — "Ravi · Dhaam Support" — is what
   * gives the customer both facts this SPECIFIC case gets from the store name
   * alone: who, and through whom.
   */
  function labelFor(session: Pick<ChatSession, 'status' | 'handledBy'> | null): string {
    if (!isGenericTitle(fallbackTitle)) return fallbackTitle;
    const name = currentAgentName(session);
    if (name === null) return fallbackTitle;
    // A host can hand this component a fallback title that already IS the
    // agent's name — the platform's own default bot name ("Assistant")
    // colliding with a placeholder widget title, most often. Pairing a
    // string with itself ("Assistant · Assistant") repeats the one fact on
    // screen twice instead of adding a second one, so the bare name wins
    // rather than the pairing.
    if (PLATFORM_BRAND.trim().toLowerCase() === name.trim().toLowerCase()) return name;
    return `${name} · ${PLATFORM_BRAND}`;
  }

  function update(session: Pick<ChatSession, 'status' | 'handledBy'> | null): void {
    const label = labelFor(session);
    node.textContent = label;

    const kind = session !== null && isHandledByCurrent(session) ? session.handledBy?.kind : undefined;
    node.setAttribute('data-handled-by', kind ?? '');

    if (!seenAnyState) {
      seenAnyState = true;
      currentLabel = label;
      return;
    }
    if (label === currentLabel) return;
    currentLabel = label;
    // The announcement stays on the bare name (or the bare fallback, with no
    // agent) — "You're now chatting with Ravi · Dhaam Support" reads like a
    // company name got appended to a person by mistake, where the visual
    // header pairing them side by side reads fine.
    liveRegion.textContent = `You're now chatting with ${currentAgentName(session) ?? fallbackTitle}.`;
  }

  function setFallbackTitle(title: string): void {
    fallbackTitle = title;
    node.textContent = title;
    currentLabel = title;
  }

  function setTitle(title: string): void {
    fallbackTitle = title;
    node.textContent = title;
    currentLabel = title;
  }

  return { node, liveRegion, update, setFallbackTitle, setTitle };
}

File: packages/flutter/lib/src/ui/header/identity_header.dart
/// Who the customer is talking to, in the panel header.
///
/// Ports `packages/widget/src/ui/identity-header.ts`.
///
/// ── The two semantics this file exists to get right ──────────────────────
///
/// [SessionSnapshot.handledBy] carries exactly one bit of presentation
/// information, and getting either half of it wrong tells the customer
/// something false:
///
///   1. ABSENT means "render your own configured title" — never "nobody is
///      handling this chat". `status`/`mode` carry that signal, not this
///      field. An absent `handledBy` must never blank the header, spin it, or
///      say "no agent". `dhaam_chat`'s own doc on that field says the same
///      thing, unprompted, because hosts kept reading it the other way.
///
///   2. PRESENT does not always mean CURRENT. A session reactivated from
///      CLOSED/RESOLVED keeps its previous agent server-side, so `handledBy`
///      can still name someone while `status` is back to `WAITING_FOR_AGENT`
///      — that person is not actually on the chat right now.
///
/// Both rules collapse into one call. [isHandledByCurrent] already returns
/// false for an absent `handledBy` (its first conjunct), so gating the whole
/// label on it handles rule 1 and rule 2 with a single branch: there is no
/// second, separate "is it absent" check to keep in sync with it.
library;

import 'package:dhaam_chat/dhaam_chat.dart';
import 'package:flutter/material.dart';

/// Whether [session]'s `handledBy` names somebody who is on the chat NOW.
///
/// The Dart counterpart of core's `isHandledByCurrent`
/// (`packages/core/src/client/session.ts:259`), ported verbatim:
///
/// ```ts
/// return session.handledBy !== undefined && session.status !== 'WAITING_FOR_AGENT';
/// ```
///
/// ── Why it lives here and not in `dhaam_chat` ────────────────────────────
///
/// `dhaam_chat` is where it belongs — it is a question about a protocol type,
/// and core is where TypeScript keeps it. It is here instead because this
/// package is its only consumer today and `dhaam_chat`'s barrel is not this
/// node's to reopen. If a second package ever needs it, move it down rather
/// than writing a second copy: two derivations of this fact is precisely the
/// stale-agent bug it exists to close.
///
/// ── Never re-derive this at a call site ──────────────────────────────────
///
/// Not `session.handledBy != null`, not a local `WAITING_FOR_AGENT` special
/// case. The title and the avatar beside it both come through here, which is
/// what keeps a face of Ada from sitting next to "Acme Support".
bool isHandledByCurrent(SessionSnapshot session) =>
    session.handledBy != null && session.status != ChatStatus.waitingForAgent;

/// The name to show, given a session and the widget's own configured title.
///
/// `session == null` is a legitimate input (no session yet, or one not
/// loaded) and always resolves to [fallbackTitle] — it is not a distinct
/// "error" state needing different handling.
String identityLabel(SessionSnapshot? session, String fallbackTitle) {
  if (session == null) return fallbackTitle;
  if (!isHandledByCurrent(session)) return fallbackTitle;
  // Read back through the object rather than force-unwrapped.
  // `isHandledByCurrent` guarantees this is non-null when it returns true,
  // but that is a compile-time guarantee about our own call sites, not a
  // runtime one about wire-sourced data.
  final HandledBy? handledBy = session.handledBy;
  return handledBy == null ? fallbackTitle : handledBy.displayName;
}

/// The kind of handler actually driving the title, or null.
///
/// A testing and styling hook for the identity DRIVING the title, not merely
/// for whether one is present: null covers both the fallback case and a stale
/// (not-current) `handledBy` alike, on purpose, since both render identical
/// copy.
HandledByKind? identityKind(SessionSnapshot? session) {
  if (session == null || !isHandledByCurrent(session)) return null;
  return session.handledBy?.kind;
}

/// The sentence the live region speaks on a real hand-off.
String identityAnnouncement(String label) => "You're now chatting with $label.";

/// The panel's title, and the announcement channel for identity CHANGES.
///
/// ── Why props rather than the reference's `update()`/`setFallbackTitle()` ─
///
/// `createIdentityHeader` is imperative: `update(session)` recomputes and may
/// speak, `setFallbackTitle(title)` recomputes and never speaks. Both rules
/// survive here, but they fall out of the props rather than out of two
/// methods a caller has to pick between:
///
///  * `setFallbackTitle` "only repaints when the fallback is what is
///    CURRENTLY on screen" — because an agent's name outranks a configured
///    title and stamping over it would rename the person the customer is
///    talking to. Here [identityLabel] simply recomputes: with an agent
///    current it returns the agent's name whatever [fallbackTitle] now says,
///    so the no-op is structural rather than a branch someone can drop.
///
///  * `setFallbackTitle` is SILENT. A screen-reader user told "You're now
///    chatting with Acme Support" because a config fetch landed would be a
///    lie about an event that did not happen. Here a [fallbackTitle] change
///    is recognised in [didUpdateWidget] and re-syncs the watermark without
///    announcing.
///
/// ── Never on the first build ─────────────────────────────────────────────
///
/// The first thing this widget renders describes whatever was ALREADY true
/// when the panel appeared — a resumed session an agent was already handling
/// — not a live hand-off. Announcing "you're now chatting with Ada" the
/// instant the widget mounts, for a fact that predates the mount, is the same
/// class of hostility as announcing forty backfilled messages on open. Same
/// discipline, and the same reason, as the message list's own first-state
/// suppression.
class IdentityHeader extends StatefulWidget {
  const IdentityHeader({
    super.key,
    required this.session,
    required this.fallbackTitle,
    this.style,
  });

  /// The session driving the identity, or null for none loaded.
  final SessionSnapshot? session;

  /// The widget's own configured title — `RemoteConfig.title`, or the host's.
  ///
  /// Shown whenever there is no CURRENT handler to name. Replacing it is the
  /// published-config path: it lands after mount and must be able to take
  /// over without announcing anything.
  final String fallbackTitle;

  final TextStyle? style;

  @override
  State<IdentityHeader> createState() => _IdentityHeaderState();
}

class _IdentityHeaderState extends State<IdentityHeader> {
  /// What is on screen right now. The watermark every announcement decision
  /// compares against — kept in step even when nothing is announced, so the
  /// NEXT change still compares against what is really displayed.
  ///
  /// Seeded in [initState], NOT as a `late` field initializer. A `late`
  /// initializer is evaluated on first READ, and the first read is inside
  /// [didUpdateWidget] — by which point `widget` is already the NEW one, so
  /// the watermark would seed itself to the value it is about to be compared
  /// against and the first real hand-off would announce nothing. Seeding here
  /// is what makes "never on the first build" mean the first build rather
  /// than the first change.
  String _currentLabel = '';

  /// The sentence to speak, or null when there is nothing to say.
  ///
  /// Cleared to null and re-set rather than held: a live region that is
  /// present the whole time is a node the platform may re-announce on an
  /// unrelated rebuild — the same rule `FormStatusLine` states for its own.
  String? _announcement;

  @override
  void initState() {
    super.initState();
    _currentLabel = identityLabel(widget.session, widget.fallbackTitle);
  }

  @override
  void didUpdateWidget(IdentityHeader oldWidget) {
    super.didUpdateWidget(oldWidget);
    final String label = identityLabel(widget.session, widget.fallbackTitle);

    // The `setFallbackTitle` path: published config landing. Silent, always —
    // see the class doc. The watermark still moves, so a genuine hand-off
    // after this still compares against what is displayed.
    if (widget.fallbackTitle != oldWidget.fallbackTitle) {
      _currentLabel = label;
      if (_announcement != null) setState(() => _announcement = null);
      return;
    }

    if (label == _currentLabel) return;
    _currentLabel = label;
    setState(() => _announcement = identityAnnouncement(label));
  }

  @override
  Widget build(BuildContext context) {
    final String label = identityLabel(widget.session, widget.fallbackTitle);
    final String? announcement = _announcement;

    return Row(
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        Flexible(
          child: Text(
            label,
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
            style: widget.style ?? Theme.of(context).textTheme.titleMedium,
          ),
        ),
        // The announcement channel, separate from the title itself and
        // carrying no visible layout — the direct port of the reference's
        // dedicated `liveRegion` node, and for its stated reason: marking the
        // title live would re-announce it on every unrelated header
        // re-render, and folding it into the message log's region would race
        // whatever that is announcing at the same moment.
        if (announcement != null)
          Semantics(
            liveRegion: true,
            label: announcement,
            child: const SizedBox.shrink(),
          ),
      ],
    );
  }
}
